Create boat class

SPDonaghy opened the pull request:
### Description
- Resolves #29 

1. A module titled obstacles.py was created
2. An obstacle Interface was created for later functionality
3. A Boat class was implemented, as well as an is_valid function
4. A test file titled test_obstacles.py was created to test the Boat init function and visualize a boat's collision cone.


### Verification

1. Tested the Boat init function by creating a boat, and verifying its dimensions, position, and orientation with a plotly scatter plot
2. Tested the is_valid function by passing points that are inside and outside of the boat's collision cone.

SPDonaghy committed:
creation of obstacles.py
```

SPDonaghy committed:
test

SPDonaghy committed:
test

SPDonaghy committed:
Set up frames for Boat and Obstacle classes

SPDonaghy committed:
Merge branch 'main' into user/SPDonaghy/29-create-boat-class

SPDonaghy committed:
Outlines and constructors for Obstacle and Boat

SPDonaghy committed:
test to check permission to push

SPDonaghy committed:
testing permission to push w/out workaround

SPDonaghy committed:
Added hull to boat attributes, added a test file

SPDonaghy committed:
Fixed a problem with arrays in obstacles.py

SPDonaghy committed:
Transposed rotation matrix

SPDonaghy committed:
added some todo notes after the sync meeting

SPDonaghy committed:
implemented a collision cone and helper functions

SPDonaghy committed:
rotations/translation work. dimensions are still off

SPDonaghy committed:
Implemented is_valid method.
Still a problem with dimensions of the collision cone

SPDonaghy committed:
Added a comment to remove ship hole from
collision cone

SPDonaghy committed:
Dimensions and shape of collision cone validated

SPDonaghy committed:
Ensured boat init func can be passed posistion as
XY type

SPDonaghy committed:
fixed axes in the plotly chart

SPDonaghy committed:
changed the is_valid function to use
Shapely contains instead of within

SPDonaghy commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,152 @@
+        x, y = position[0], position[1]
+
+        # Points of the boat collision box polygon before rotation and centred at the origin
+        points = np.array(
```
I would like to receive feedback on how to shape the boat's collision box.  I am defining the collision box as the region where the boat may possibly be, over a certain amount of time: delta_time. 

This array contains the set of points which describe the boat's collision box.

Right now, it is shaped like a trapezoid. with the narrower end being where the boat currently is, and the wider end extends a certain distance ahead depending on the boat's length, speed, and delta-time (which are all parameters passed to the Boat init function.

I chose this shape so that, if the boat is not moving, its hull is still completely contained in the collision box.

But I am also considering defining the collision box as follows: 

The ship is inscribed in a circle, and a cone shape extends from it, in the direction of the course over ground (track made good). This shape will ensure that even if the ship's heading does not match its COG then no points on the boat will stick out of the collision box. This can happen with the simpler trapezoid shape

SPDonaghy commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,152 @@
+from shapely.geometry import Polygon
+
+
+# TODO Look into adding more data available from AIS
```
Rate of turn might be a useful data value to better depict the expected path of boats

SPDonaghy commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,152 @@
+            width, length, position, speed, delta_time, course_over_ground
+        )
+
+    def create_collision_cone(
```
Later on, I would like to see how I can reuse this function (or the entire class) to create land objects from geo data.

SPDonaghy commented on test/test_obstacles.py:
```diff
@@ -0,0 +1,29 @@
+import numpy as np
+import plotly.graph_objects as go
```
I think Plotly will be essential in visualizing what I am working on with obstacles, moving forward. However, it runs very slowly sometimes, I tried the commands for setting the hostname and that does improve it, but I am finding I need to do that every time I start the codespace.

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,152 @@
+
+    """
+
+    def __init__(self, id, width, length, position, speed, delta_time, course_over_ground):
```
Arguments should include https://github.com/UBCSailbot/custom_interfaces/blob/main/msg/HelperAISShip.msg and any other information needed. Thus latlon to xy conversion should be done inside this, and you will need to pass in the reference point (next global waypoint)

Example of importing msgs in codebase:
https://github.com/UBCSailbot/local_pathfinding/blob/a3dc649ab4cc077ed3a4a0e7ef7b23b93d8b8a92/local_pathfinding/node_navigate.py#L4

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,152 @@
+            -length (float): length of the boat in meters
+            -position (XY): x,y coordinates of the boat converted from lat/lon
+            -speed (float): speed of the boat in knots
+            -delta_time (float): time in seconds, to calculate the boat's expected position
```
Rather than time it should use the current position and maximum speed of sailbot

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,152 @@
+        Returns:
+            bool: True if the point is not within the boat's collision cone
+        """
+        return not self.collision_cone.contains(point)
```
This should be able to be defined inside `ObstacleInterface`

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,152 @@
+
+# TODO Look into adding more data available from AIS
+# TODO Decide between using COG and SOG or Heading and Speed
+class ObstacleInterface:
```
Rename to `Obstacle`

patrick-5546 commented on test/test_obstacles.py:
```diff
@@ -0,0 +1,29 @@
+
+# check the is_valid function
+assert not boat1.is_valid(Point(100, 50))
+assert boat1.is_valid(Point(100, 100))
```
Wrap in test method and parameterize using pytest:
- Example in codebase
https://github.com/UBCSailbot/local_pathfinding/blob/a3dc649ab4cc077ed3a4a0e7ef7b23b93d8b8a92/test/test_ompl_path.py#L46-L60
- Pytest documentation: https://docs.pytest.org/en/7.4.x/how-to/parametrize.html

patrick-5546 commented on test/test_obstacles.py:
```diff
@@ -0,0 +1,29 @@
+x = np.array(boat1.collision_cone.exterior.coords.xy[0])
+y = np.array(boat1.collision_cone.exterior.coords.xy[1])
+
+# Create a Plotly figure for boat1's collision cone with a hole representing the hull
+fig1 = go.Figure(go.Scatter(x=x, y=y, fill="toself"))
+
+fig1.update_layout(yaxis_range=[-200, 200], xaxis_range=[-200, 750])
```
Put the plotly stuff (manual inspection) at the end and only run when running the file
```py
if __name__ == "__main__":
    import plotly.graph_objects as go
    ...
```

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,152 @@
+        )
+
+    def create_collision_cone(
+        self, width, length, position, speed, delta_time, course_over_ground
```
Add type hints to all functions

patrick-5546 reviewed (commented):
Great start! Here are a couple things I noticed:

SPDonaghy committed:
modified the args to Boat init
to accept ais_ship object
modified how the collision zone is shaped
for a Boat

SPDonaghy committed:
modified is_valid to take LatLon as argument
instead of Shapely Point

SPDonaghy committed:
changed COG and SOG to cog and sog
modified test_obstacles.py
but it still doesn't run because
custom_interfaces doesn't rebuild properly

SPDonaghy committed:
Improvements to test file
modified function defs in obstacles
added a meters_to_km function to coord_systems
calc intersect time does not work when
sailbot can catch up to the boat
error is with np.roots

SPDonaghy committed:
Corrected error in argument for np.roots()

SPDonaghy committed:
Added states to the test plot
changed some internal documentation
Added ROT to ais_ship for later use

SPDonaghy committed:
Moved some functions to the boat class
added sailbot speed arg to boat class
changes to make mypy happy

SPDonaghy committed:
Added a safety buffer to collision cone
Updated test file to check is_valid and to
visually check the collision cone
Added links to more in depth explanations to
some functions and tests

SPDonaghy commented:
I believe it is just failing colcon because my changes to custom_interfaces are not merged to main. I have submitted a PR for those changes.

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,340 @@
+
+    The choice was made to create an interface rather than a class as Boat and LandMass objects
+    may be very different objects which share an "avoidable" property and a position, but not much
+    else.
```
I don't think this is true anymore

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,340 @@
+        Args:
+            -reference (LatLon): latitude and longitude of the next global waypoint
+            -sailbot_position (LatLon): lat and lon position of SailBot
+            -sailbot_speed (float): speed of the SailBot in km/h
```
I don't think we need a docstring for the init function in this case. However, let's move this section to the class attributes.

For example
https://github.com/UBCSailbot/local_pathfinding/blob/bd5a710ba1af486d615790a4b2f9c7f94dc87372/local_pathfinding/local_path.py#L53-L61

patrick-5546 replied:
The only difference is that `sailbot_position` would be type `XY` instead of `LatLon`

SPDonaghy replied:
So sailbot_position will be published as an XY correct?
Will reference also be published as XY? 

I'm a little confused on what will be XY and what will be LatLon, I assumed I needed to use latlon_to_xy in my functions, because locations will be published in LatLon, passed in to functions as LatLon, and then converted to XY inside my functions.

patrick-5546 replied:
Hopefully we can keep the XY coordinate system internal to this file (not thought about anywhere else). The `self.sailbot_position` class attribute would be type XY. The reference XY coordinate would be (0, 0)

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,340 @@
+            -length (float): length of the boat in meters
+    Notes
+    ----------
+        The following information about a boat is obtained from AIS:
+            - Maritime Mobile Service Identity (9 digit int)
+            - latitude (float) latitude in degrees
+            - longitude (float) longitude in degrees
+            - speed over ground (float) speed in knots over ground
+            - course over ground (float) degrees measured clockwise from true north
+            - boat dimensions (float) width and length of the boat in meters
+            - rate of turn (float) ROT in AISROT scale -126 to +126 corresponding to
+                -708 to +708 deg/min
+
+            - other information is also available but not used here
```
Scope: shouldn't be documented here. Docstrings should only include what is necessary to understand the implementation. Can you move this to `HelperAISShip.msg`

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,340 @@
+            -sailbot_position (LatLon): lat and lon position of SailBot
+            -sailbot_speed (float): speed of the SailBot in km/h
+            -ais_ship (HelperAISShip): an AISShip object, containing the following information:
+                -id (int): MMSI number of the boat
+                -width (float): width of the boat in meters
+                -length (float): length of the boat in meters
+                -lat_lon (LatLon): latitude and longitude of the boat
+                -speed_over_ground (float): speed of the boat in knots, over ground
+                -course_over_ground (float): COG of the boat, in degrees, clockwise from true north
+                -rate_of_turn (float): ROT of the boat in AISROT scale -126 to +126 corresponding
+                    to -708 to +708 degrees per minute
```
Again no need to document here, people can refer to the definition of `HelperAISShip.msg`

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,340 @@
+    ----------
+        -id (int): MMSI number of the boat
+        -speed_over_ground (float): speed of the boat, over ground, in knots
+        -collision_cone (Polygon): a shapely Polygon object representing the boat's collision box
```
This should contain only attributes that are unique to `Boat`. Should move `collision_cone` to `Obstacle` attributes

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,340 @@
+        sailbot_speed: float,
+        ais_ship: HelperAISShip,
+    ):
+        """
```
imo no need to add docstring to init function unless you're doing something crazy

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,340 @@
+        The polygon is oversized according to the collision zone safety buffer, for
+        added assurrance that the boat will be entirely contained by the polygon.
+
+        Args:
```
Missing `sailbot_speed`

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,340 @@
+
+        # This factor can be adjusted to change the scope/width of the collision cone
+        # TODO This feels too arbitrary, maybe will incorporate ROT at a later time
+        COLLISION_CONE_STRETCH_FACTOR = projected_distance * 1.5
```
All caps reserved for constants defined at the top of the file. Can make `1.5` a constant

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,340 @@
+            position (XY): x,y coordinates of the boat in km
+            speed_over_ground (float): speed of the boat in knots, over ground
+            course_over_ground (float): COG of the boat in degrees clockwise from true north
+            rate_of_turn (float): ROT of the boat in AISROT scale -126 to +126 corresponding to
```
rate of turn unused

patrick-5546 replied:
Can remove it until/if it is used?

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,340 @@
+        t = np.roots(quadratic_coefficients)
+
+        # Return the smaller positive time
+        return min([i for i in t if i > 0])
```
Is it possible that there could be no element > 0? If so, `min` would raise a `ValueError`

SPDonaghy replied:
Yes, that is possible. I will add a try except block to check for that and return -1 if there are no positive elements. 

Generally, its ok if there doesn't end up being any positive times, just means they cant collide with the current parameters. So projecting the Boat's position up to the maximum distance would be the appropriate action in that case.

patrick-5546 committed:
Formatting

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,340 @@
+        self.sailbot_speed = sailbot_speed
+
+        # This should be defined in child classes
+        self.collision_zone = Polygon(None, None)
```
What do you think about this for debugging purposes
- Set `self.collision_zone = None` here
- In `is_valid()`
  ```py
  if self.collision_zone is None:
    raise ValueError("Collision zone has not been initialized")
  ```

patrick-5546 replied:
If you implement this, you will probably want to add the corresponding test

SPDonaghy replied:
I like this idea, I will implement it.

patrick-5546 replied:
You will also want to update the docstring

```py
"""

Raises:
    ValueError: When collision zone has not been initialized.
"""
```

patrick-5546 reviewed (commented):
Great work on this PR! I am really impressed by your code quality and readability. I aligned the docstring formatting with the rest of the codebase and noted a couple small things.

patrick-5546 committed:
More formatting

SPDonaghy committed:
Implemented bulk of changes suggested in PR review

SPDonaghy committed:
small changes to docstrings
final prob is with the roots in calc proj time
still need to solve

SPDonaghy committed:
resolved potential for ValueError in
calc_projected_time() with try/except

SPDonaghy committed:
implemented more changes from PR review
added checks for collision_zone being None
improved check for positive intersection to be more
specific than just catching any ValueError

Fixed a bug in the test file, that arose from knots and kmph
conversions

SPDonaghy committed:
just changed a couple words

SPDonaghy committed:
Changed is_valid to just use self.reference
internally, instead of passing in reference as
a parameter.

SPDonaghy committed:
now pass XY to is_valid instead of LatLon
Based on input from Sync meeting

patrick-5546 committed:
Nitpicks

patrick-5546 reviewed (approved):
I think this PR is ready to merge, great job! You'll have to merge in your custom_interfaces PR first and rerun failed tests.

Notes about my nitpicks:
- Add docstring for any exceptions raised
- `collision_zone` is defined in `Obstacle`, so to respect that in `Boat` I call an Obstacle function that updates the collision zone
- Class attributes are any instance variables you define

What would be good to work on next is using this in the other files to add obstacle avoidance.

patrick-5546 commented:
~~Also it would be great if you presented your work at the meeting on Saturday to show everyone what you've been up to!~~ actually not done yet lol

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,267 @@
+
+        collision_zone = self.create_collision_cone(
+            ais_ship.dimensions.width,
+            ais_ship.dimensions.length,
```
You can store all these arguments as instance variables so that you don't have to pass them to each function

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,267 @@
+            ais_ship.sog.sog,
+            ais_ship.cog.cog,
+            self.sailbot_position,
+            self.sailbot_speed,
```
Use the instance variables inside the function definition rather than passing it in as an argument

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,267 @@
+
+        # Calculate distance the boat will travel before soonest possible collision with Sailbot
+        projected_distance = self.calculate_projected_distance(
+            position,
+            course_over_ground,
+            speed_over_ground,
+            sailbot_position,
+            sailbot_speed,
```
Same for these

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,267 @@
+        )
+
+        time_to_intersection = self.calculate_time_to_intersection(
+            position, boat_sog_vector, sailbot_position, sailbot_speed
```
Same for these

patrick-5546 commented on test/test_obstacles.py:
Do you think it would be possible to add more unit tests for the boat obstacle? Like tests for create_collision_cone, calculate_projected_distance, and calculate_time_to_intersection

patrick-5546 replied:
If possible the test arguments should make sense intuitively (not always require calculation). If it does require calculation it would be helpful to add the function that calculates it.

patrick-5546 commented on local_pathfinding/coord_systems.py:
```diff
@@ -45,6 +45,22 @@ def cartesian_to_true_bearing(cartesian: float) -> float:
     return (90 - cartesian + 360) % 360
 
 
+def knots_to_kilometers_per_hour(knots: float) -> float:
```
Although trivial, we should add tests for each function

patrick-5546 replied:
Also is this function even used anywhere?

SPDonaghy replied:
No, it used to be. But if speeds can be published in kmph its not needed

patrick-5546 reviewed (changes requested):
Sorry thought of some more things to do in this PR

SPDonaghy committed:
Made changes according to PR review.
main change is using instance variables to create collision zone
instead of passing variables as args

SPDonaghy commented:
**dynamic variables:** 
- sog
- cog
- position
- sailbot_position
- sailbot_speed

**non-dynamic variables:**
- id
- width
- length

Hi Patrick, I am just resolving some conflicts, with my most recent changes and your commit. (Lesson learned I should have pulled before making more changes)

Having update_collision_zone(Polygon) as a non static method makes sense to me. It seems like the intent is that we can just call this method to update the dynamic variables instead of making a new boat object every time something changes.

I am just wondering for create_collision_cone() and its helper functions (proj distance and proj time) do we want these to be non-static as well? Is it okay for them to still have parameters for the dynamic variables listed above? And then just use the non dynamic instance variables (self.width and self.length) inside these functions?

I interpreted your previous comments about using instance variables to mean we shouldn't pass anything to these functions as parameters. But that would require storing the dynamic variables as instance variables. Which is something I want to avoid, so that we don't have that data stored in the instance and in the instance's collision_zone.

patrick-5546 commented:
> Lesson learned I should have pulled before making more changes

Yes this is something we all needed to go through, but glad you did sooner rather than later lol

> I am just wondering for create_collision_cone() and its helper functions (proj distance and proj time) do we want these to be non-static as well? Is it okay for them to still have parameters for the dynamic variables listed above? And then just use the non dynamic instance variables (self.width and self.length) inside these functions?

You make a good point. Rather than a direct answer, I will give some more background about the usecases of the Boat class. Primarily it is to feed into our path planner for collision avoidance. But our second usecase is to visualize boat's and their predicted positions on the website. Thus we will want to ensure that the collision zone that we are publishing was computed using the dynamic information that we are publishing. One way to do this is to store the dynamic variables as instance variables. along with the collision zone. Side note: it might be helpful to have the time of the information as well.

For `create_collision_cone()`, we can pass in an argument of type `HelperAISShip`. Then it would update the dynamic instance variables and create the polygon. May be good to raise an exception if the id of the argument doesn't match `self.id`. Also infrastructure wise, one idea is to overload `create_collision_zone()`: rename `create_collision_cone()` to `create_collision_zone(HelperAISShip)` and make its last line `self.create_collision_zone(Polygon)`.

For proj distance and proj time, we could make these class variables (does not use instance variables, `@staticmethod`). This would make them easier to test separately.

> But that would require storing the dynamic variables as instance variables. Which is something I want to avoid, so that we don't have that data stored in the instance and in the instance's collision_zone.

Not sure if I understand this, how are the dynamic variables stored in the collision zone?

SPDonaghy commented:
Thanks for the comments! I'll get working on this again tonight or tomorrow. 

> Not sure if I understand this, how are the dynamic variables stored in the collision zone?

No you're correct, they are used to make the collision zone, not really stored in it.

SPDonaghy committed:
Merge branch 'user/SPDonaghy/29-create-boat-class' of https://github.com/UBCSailbot/local_pathfinding into user/SPDonaghy/29-create-boat-class

SPDonaghy committed:
small changes to docstrings and function args

SPDonaghy committed:
removed overloading

SPDonaghy committed:
Merge branch 'main', remote-tracking branch 'origin' into user/SPDonaghy/29-create-boat-class

SPDonaghy committed:
implemented more tests
added auto calculation to visual test

SPDonaghy committed:
changed how the collision zone is measured

SPDonaghy committed:
added coverage for case where proj dist should be 0

SPDonaghy commented:
I realized, along with functions for updating a boat's collision zone periodically, The Obstacle class needs a function to update dynamic information about Sailbot (position and speed) periodically as well. I'll write something up later today.

Also I saw your comment about including timestamps in AISShip messages as well. As I understand, it won't really be needed in this module for anything. And maybe I can just make a new pr in custom_interfaces at some point, to add that?

Also, I still want to explore how missing/unavailable data should be handled. (Ex. I've seen A COG of 501 means unavailable)

SPDonaghy committed:
added tests for added update functions

patrick-5546 commented:
> Also, I still want to explore how missing/unavailable data should be handled. (Ex. I've seen A COG of 501 means unavailable)

This should be handled by NET, @hhenry01 FYI

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,254 @@
+
+        return not self.collision_zone.contains(point)
+
+    def update_collision_zone(self, collision_zone: Polygon):
```
What do you think about adding what is in `create_collision_zone` here and removing it? This would reduce the number of time we go back and forth between `Boat` and `Obstacle`

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,254 @@
+    Also referred to target ships or boat obstacles.
+
+    Attributes:
+        id (int): MMSI ID of the boat.
+        width (float): Width of the boat in meters.
+        length (float): Length of the boat in meters.
```
Only attribute now is `ais_ship`

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,254 @@
+        self.ais_ship = ais_ship
+        self.update_collision_zone(self.create_boat_collision_zone(self.ais_ship))
+
+    def create_boat_collision_zone(self, ais_ship: HelperAISShip) -> Polygon:
```
How about we make this `update_boat_collision_zone`? Then we can call it directly in init. The last line of this function would call `self.update_collision_zone`

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,254 @@
+
+        # Ensure the instance variable ais_ship is the most up-to-date version
+        # Because this function may be called outside of the constructor
+        self.ais_ship = ais_ship
```
One way to resolve this redundancy is to make the `ais_ship` argument optional

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,254 @@
+
+        return Obstacle.create_collision_zone(collision_zone_poly, position)
+
+    def calculate_projected_distance(self, ais_ship: HelperAISShip) -> float:
```
If you pass in the boat attributes needed directly you can make this static

patrick-5546 replied:
Why not add implementation in `calculate_time_to_intersection` here? This function is pretty small, but ultimately utu just a suggestion

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,254 @@
+
+        return t * ais_ship.sog.speed
+
+    def calculate_time_to_intersection(self, ais_ship: HelperAISShip) -> float:
```
If you pass in the boat attributes needed directly you can make this static

patrick-5546 commented on local_pathfinding/obstacles.py:
```diff
@@ -0,0 +1,254 @@
+
+        if isinstance(self, Boat):
+            # regenerate collision zone with updated reference point
+            self.update_collision_zone(self.create_boat_collision_zone(self.ais_ship))
```
Can replace with just `self.update_boat_collision_zone` if following suggestions above

patrick-5546 reviewed (commented):
I really like the changes you made! I think these suggestions should be the last things from me.

patrick-5546 commented on local_pathfinding/test.txt:
Was this committed by mistake?

SPDonaghy committed:
consolidated create_collision_zone and
update_collision_zone
consolidated calc_proj_dist and
calc_time_to_intersect
made calc_proj_dist static
updated tests to reflect changes

SPDonaghy committed:
test.txt is comitted by mistake

patrick-5546 committed:
Fix type hint and undo staticmethod

patrick-5546 reviewed (approved):
Looks good to me. Would be interested in diving deeper into the math at some point, perhaps in a SW design review.

I changed `calculate_projected_distance()` back to an instance function because I saw that all the arguments that it was being passed were instance variables.

Also for performance reasons we would like to do as little math in raw Python as possible (within reason), instead using `math` for single numbers and `numpy` for arrays. Can any of the raw math that you did in `update_collision_zone()` or `calculate_projected_distance()` be converted to use these libraries? And for calculations that are performed multiple times, consider using a variable

patrick-5546 commented on test/test_obstacles.py:
```diff
@@ -0,0 +1,463 @@
+
+
+if __name__ == "__main__":
+    """VISUAL TESTS
+
+    The collision zone length can be verified visually, using the plotly chart below.
+    """
```
Would be helpful to add what to check to verify that everything is working

SPDonaghy committed:
Optimization improvements
Added what to look for in visual tests

SPDonaghy commented:
I made a few optimization improvements:

-  storing repeatedly calculated/retrieved values in variables
- importing specific functions from libraries to avoid using dot operator or importing entire libraries
- reduced the number of polygon objects created during updates by combining rotation and translation into a single transform.

I wasn't sure though, how to incorporate math to make any of my calculations not run in python. Would you be able to explain more what you mean by using math for single numbers?

patrick-5546 committed:
Import entire computation libraries

patrick-5546 committed:
Revert instance -> local variables

patrick-5546 committed:
Remove unneeded else branch

patrick-5546 commented:
I think what you have is good. I reverted some changes because although they're technically faster, they aren't a bottleneck and make the code less readable and concise
